aws/validators.go

```diff
@@ -977,6 +977,20 @@ func validateLogGroupNamePrefix(v interface{}, k string) (ws []string, errors []
 	return
 }
 
+// https://docs.aws.amazon.com/AmazonS3/latest/userguide/bucketnamingrules.html
+func validateS3BucketName() schema.SchemaValidateFunc {
+	// skipped due to ease of implementation:
+	// - must not be formatted as an IP(v4?) address
+	return validation.All(
+		// length
+		validation.StringLenBetween(3, 63),
+		// allowable characters, and alphanumeric start + end
+		validation.StringMatch(regexp.MustCompile(`^[a-z0-9][a-z0-9.-]+[a-z0-9]$`), ""),
```

Comment:
This can actually all be in a single regex. I'm using it in a custom s3 bucket module that defaults a few values and it works well:
```
(?=^.{3,63}$)(?!^(\\d+\\.)+\\d+$)(^(([a-z0-9]|[a-z0-9][a-z0-9\\-]*[a-z0-9])\\.)*([a-z0-9]|[a-z0-9][a-z0-9\\-]*[a-z0-9])$)
```

Reference - https://stackoverflow.com/a/50484916